### PR TITLE
Lazy-loading support

### PR DESCRIPTION
**Pull request by kripken**
This adds `emscripten_lazy_load_code()`, a function that when called will block (using Asyncify) and load the complete program. This lets the initial download not contain code that can be proven to be used only after those calls.

How this works is we emit the initial downloaded wasm after modifying it with Binaryen to assume that those calls will not rewind, as we only unwind using that binary (the optimizer can then remove a lot of code). We also emit the later downloaded wasm, optimized the other way, to assume we only rewind but never unwind. These transforms are done using ModAsyncify passes from Binaryen https://github.com/WebAssembly/binaryen/pull/2404

A new option `ASYNCIFY_LAZY_LOAD_CODE` is necessary to use this (as well as `ASYNCIFY` itself).

The test shows an example where the initial download is less than half the size of the later download. Note that in this model we do download some code twice (the later download contains the code in the initial one), and we do add some code size due to Asyncify support. So this will not be an obvious win for every codebase. But if the later download contains unlikely code that may never be used, and there is a lot of such code, and we can ignore indirect calls for Asyncify's purposes, then the win can be significant.

cc @surma @RReverser

### CODE REVIEW

File: tools/shared.py [hunk 1]
```diff
@@ -2853,6 +2871,18 @@ def get_binaryen_bin():
...
+  @staticmethod
+  def get_binaryen_command(tool, args=[], debug=False):
+    cmd = [os.path.join(Building.get_binaryen_bin(), tool)] + args
+    cmd += Building.get_binaryen_feature_flags()
+    if debug:
+      cmd += ['-g'] # preserve the debug info
+    return cmd
+
+  @staticmethod
+  def get_wasm_opt_command(args=[], debug=False):
```
Comment (sbc100) [inline tools/shared.py]: can you have this take and infile and outfile argument since all the call sites want that I think the command requires it?

File: tools/shared.py [hunk 2]
```diff
@@ -2538,7 +2538,7 @@ def minify_wasm_js(js_file, wasm_file, expensive_optimizations, minify_whitespac
         # If we are building with DECLARE_ASM_MODULE_EXPORTS=0, we must *not* minify the exports from the wasm module, since in DECLARE_ASM_MODULE_EXPORTS=0 mode, the code that
         # reads out the exports is compacted by design that it does not have a chance to unminify the functions. If we are building with DECLARE_ASM_MODULE_EXPORTS=1, we might
         # as well minify wasm exports to regain some of the code size loss that setting DECLARE_ASM_MODULE_EXPORTS=1 caused.
-        if not Settings.STANDALONE_WASM and not Settings.AUTODEBUG and not Settings.ASSERTIONS and not Settings.SIDE_MODULE and emitting_js:
+        if not Settings.STANDALONE_WASM and not Settings.AUTODEBUG and not Settings.ASSERTIONS and not Settings.SIDE_MODULE and emitting_js and not Settings.ASYNCIFY_LAZY_LOAD_CODE:
           js_file = Building.minify_wasm_imports_and_exports(js_file, wasm_file, minify_whitespace=minify_whitespace, minify_exports=Settings.DECLARE_ASM_MODULE_EXPORTS, debug_info=debug_info)
```
Comment (sbc100) [inline tools/shared.py]: Why can't we minify with lazy loading?  Is that something that can be enabled as a followup?

File: tests/test_core.py [hunk 3]
```diff
@@ -7709,6 +7709,89 @@ def test_emscripten_scan_registers(self):
     self.set_setting('ASYNCIFY', 1)
     self.do_run_in_out_file_test('tests', 'core', 'emscripten_scan_registers')
 
+  @no_fastcomp('wasm-backend specific feature')
+  @no_wasm2js('TODO: lazy loading in wasm2js')
+  @parameterized({
+    'conditional': (True,),
+    'unconditional': (False,),
+  })
+  def test_emscripten_lazy_load_code(self, conditional):
```
Comment (sbc100) [inline tests/test_core.py]: How about just "test_lazy_load_code"?

File: tests/core/emscripten_lazy_load_code.txt [hunk 4]
```diff
@@ -0,0 +1,2 @@
+foo_start
```
Comment (sbc100) [inline tests/core/emscripten_lazy_load_code.txt]: Can you call these files "test_lazy_load_code.*" to match the test name

File: tests/test_core.py [hunk 5]
```diff
@@ -7709,6 +7709,89 @@ def test_emscripten_scan_registers(self):
...
+    print('first wasm size', first_size)
+    print('second wasm size', second_size)
+    if not conditional and is_optimizing(self.emcc_args):
+      # If the call to lazy-load is unconditional, then the optimizer can dce
+      # out more than half
+      self.assertLess(first_size, 0.5 * second_size)
+
+    with open('src.cpp.o.wasm', 'rb') as f:
+      with open('src.cpp.o.wasm.lazy.wasm', 'rb') as g:
+        assert f.read() != g.read(), 'the lazy-loaded wasm is different'
```
Comment (sbc100) [inline tests/test_core.py]: assertNotEqual?

File: tests/test_core.py [hunk 6]
```diff
@@ -7709,6 +7709,89 @@ def test_emscripten_scan_registers(self):
...
+
+    # attempts to "break" the wasm by adding an unreachable in $foo_end. returns whether we found it.
+    def break_wasm(name):
+      wat = run_process([os.path.join(Building.get_binaryen_bin(), 'wasm-dis'), name], stdout=PIPE).stdout
```
Comment (sbc100) [inline tests/test_core.py]: get_binaryen_command? here and below.

File: tests/test_core.py [hunk 7]
```diff
@@ -7709,6 +7709,89 @@ def test_emscripten_scan_registers(self):
...
+        f.write(wat)
+      shutil.move(name, name + '.orig')
+      run_process([os.path.join(Building.get_binaryen_bin(), 'wasm-as'), 'wat.wat', '-o', name, '-g'])
+      return True
+
+    def verify_working(args=['0']):
+      self.assertContained('foo_end', run_js('src.cpp.o.js', args=args))
+
+    def verify_broken(args=['0']):
+      self.assertNotContained('foo_end', run_js('src.cpp.o.js', args=args, stderr=STDOUT, assert_returncode=None))
```
Comment (sbc100) [inline tests/test_core.py]: Use `self.expect_fail`?

File: tests/test_core.py [hunk 8]
```diff
@@ -7709,6 +7709,89 @@ def test_emscripten_scan_registers(self):
...
+
+    # restore
+    shutil.copyfile('src.cpp.o.wasm.orig', 'src.cpp.o.wasm')
+    shutil.copyfile('src.cpp.o.wasm.lazy.wasm.orig', 'src.cpp.o.wasm.lazy.wasm')
+    verify_working()
+
+    if conditional:
+      # if we do not call the lazy load function, then we do not need the lazy wasm,
+      # and we do the second call in the first wasm
+      os.unlink('src.cpp.o.wasm.lazy.wasm')
```
Comment (sbc100) [inline tests/test_core.py]: `os.remove` is a little more recognizable maybe?

File: tests/core/emscripten_lazy_load_code.cpp [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+#include <stdlib.h>
+
+extern "C" {
+
+EM_JS(void, log_started, (), {
+  out("foo_start");
+})
+
+EM_JS(void, log_ended, (), {
+  out("foo_end");
```
Comment (sbc100) [inline tests/core/emscripten_lazy_load_code.cpp]: Why not simply use puts/printf here?

File: tests/core/emscripten_lazy_load_code.cpp [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+void foo_start(int n) {
+  log_started();
+  // prevent inlining
+  if (n > 10000) foo_start(n - 1);
+  if (n > 15000) foo_start(n - 3);
+}
+
+void foo_end(int n) {
+  log_ended();
+  // prevent inlining
```
Comment (sbc100) [inline tests/core/emscripten_lazy_load_code.cpp]: How about `__attribute__((noinline))` to be explicit?

File: tests/core/emscripten_lazy_load_code.cpp [hunk 11]
```diff
@@ -0,0 +1,43 @@
+#include <emscripten.h>
```
Comment (sbc100) [inline tests/core/emscripten_lazy_load_code.cpp]: Add copyright

File: tools/shared.py [hunk 12]
```diff
@@ -2635,14 +2632,37 @@ def metadce(js_file, wasm_file, minify_whitespace, debug_info):
...
     return Building.acorn_optimizer(js_file, passes, extra_info=json.dumps(extra_info))
 
+  @staticmethod
+  def asyncify_lazy_load_code(wasm_binary_target, options, debug):
+    # create the lazy-loaded wasm. remove the memory segments from it, as memory
+    # segments have already been applied by the initial wasm, and apply the knowledge
+    # that it will only rewind, after which optimizations can remove some code
+    cmd = Building.get_wasm_opt_command(debug=debug)
+    cmd += [wasm_binary_target, '-o', wasm_binary_target + '.lazy.wasm']
+    cmd += ['--remove-memory']
```
Comment (sbc100) [inline tools/shared.py]: Why this?

File: tools/shared.py (hunk 1 above)
Comment (kripken) [reply tools/shared.py]: Sounds goos, but perhaps as a followup? That's going to be a bunch more changes not directly relevant to this PR.

File: tools/shared.py (hunk 2 above)
Comment (kripken) [reply tools/shared.py]: ASYNCIFY_LAZY_LOAD_CODE disables minification because it runs after the wasm is completely finalized, and we need to be able to still identify import names at that time. To avoid that, we would need to keep a mapping of the names and send that to binaryen. That's possible, but not trivial. I'm adding a comment to clarify.

File: tests/test_core.py (hunk 3 above)
Comment (kripken) [reply tests/test_core.py]: It's the name of the C function, though?

File: tests/test_core.py (hunk 6 above)
Comment (kripken) [reply tests/test_core.py]: wasm-dis and as don't support feature flags, so that utility function doesn't work on them. But given they don't need those flags, it's about the same length to type either way.

File: tests/test_core.py (hunk 7 above)
Comment (kripken) [reply tests/test_core.py]: `expect_fail` takes a console command to run, so it's not compatible with `run_js` which hides the specific command. Maybe we can think of a new API like `expect_fail_js`?

File: tests/core/emscripten_lazy_load_code.cpp (hunk 9 above)
Comment (kripken) [reply tests/core/emscripten_lazy_load_code.cpp]: stdio increases code size, and it's nice in the test to only use it after the lazy loading, to see the size difference more clearly.

File: tests/core/emscripten_lazy_load_code.cpp (hunk 10 above)
Comment (kripken) [reply tests/core/emscripten_lazy_load_code.cpp]: That's not enough for binaryen, though :(

File: tools/shared.py (hunk 12 above)
Comment (kripken) [reply tools/shared.py]: The memory? See the comment a few lines up, `remove the memory segments from it, as memory segments have already been applied by the initial wasm`

File: tests/test_core.py (hunk 3 above)
Comment (sbc100) [reply tests/test_core.py]: OK SGTM either way.     I was think thinking "lazy_load_code" is kind of concept you are testing since its also the suffix of `ASYNCIFY_LAZY_LOAD_CODE`.

File: tools/shared.py (hunk 1 above)
Comment (sbc100) [reply tools/shared.py]: How about splitting the `get_wasm_opt_command` part out of this change?  Seems like a refactor that would make this change smaller.  I leave that up to you though, I know you wan to get this landed today if possible so i don't want to delay you too much.

Comment (kripken) [reply tools/shared.py]: Well, the reason I added it here is that it made this shorter to write, and also seemed like a useful change. I don't see a benefit to splitting it out, but I agree this can be improved more, I'll look at that as a followup.

File: site/source/docs/api_reference/emscripten.h.rst [hunk 13]
```diff
@@ -1286,10 +1288,10 @@ IndexedDB
   :param perror: An out parameter that will be filled with a non-zero value if an error occurred.
 
 
-Asyncify functions
-==================
+Fastcomp Asyncify functions
```
Comment (RReverser) [inline site/source/docs/api_reference/emscripten.h.rst]: Why was this changed to say "Fastcomp"? AFAIK same applies to Asyncify transform on top of wasm backend?

Comment (kripken) [reply site/source/docs/api_reference/emscripten.h.rst]: This PR refactored these docs, and now this part has the fastcomp funcs, while lower down there is "Upstream Asyncify".

(Or maybe I didn't understand what you meant by "same" in that sentence?)

Comment (RReverser) [reply site/source/docs/api_reference/emscripten.h.rst]: By "same" I mean that the description in the first sentence (both `-s ASYNCIFY=1` as well as link to Asyncify docs) seems generic enough, but only included in Fastcomp section now.

Also, aren't the functions that are currently included in "Fastcomp functions" - e.g. `emscripten_sleep` - available with upstream backend as well?

If they're, then the current structure seems confusing, as it seems to imply that these functions are Fastcomp-only, while others are upstream-only. I'd suggest turning the first section into "Asyncify functions" and then nesting the other one under it as "Upstream-only Asyncify functions".

Comment (kripken) [reply site/source/docs/api_reference/emscripten.h.rst]: Thanks! Now I see what you mean. Yeah, `sleep` looks wrong - it's mentioned in fastcomp, as you say, but should only be in the shared section earlier. I also see sleep with yield is in the wrong place. I opened #9819 now for these issues.

**Comment by gabrielcuvillier** [comment]
@kripken A question on this interesting feature: does this work as expected if the program have dynamically allocated memory (malloc, new, etc...) ?

I mean, suppose the program have instantiated classes with virtual methods, and one of the virtual method is doing lazy_load(). When this function is called and the lazy load occurs, it seems the program is being "replaced" by the newer version: but does the memory follows and everything is "remapped" accordingly ?

Not sure if I am clear :)

**Comment by kripken** [comment]
Yes, memory is preserved exactly as it was before, so dynamic allocation is fine.

We replace the code, but do not replace the data in memory. So we do not apply the memory initialization again, for example.
